samples/cpp/geometry.cpp

```diff
@@ -0,0 +1,252 @@
+/*******************************************************************************
+ *
+ * This program demonstrates various shape fitting techniques using OpenCV.
+ * It reads an image, applies binary thresholding, and then detects contours.
+ *
+ * For each contour, it fits and draws several geometric shapes including
+ * convex hulls, minimum enclosing circles, rectangles, triangles, and ellipses
+ * using different fitting methods:
+ *  1: OpenCV's original method fitEllipse which implements Fitzgibbon 1995 method.
+ *  2: The Approximate Mean Square (AMS) method fitEllipseAMS  proposed by Taubin 1991
+ *  3: The Direct least square (Direct) method fitEllipseDirect proposed by Fitzgibbon1999
+ *
+ * The results are displayed with unique colors
+ * for each shape and fitting method for clear differentiation.
+ *
+ *
+ *********************************************************************************/
+
+#include "opencv2/imgproc.hpp"
+#include "opencv2/highgui.hpp"
+#include <iostream>
+#include <vector>
+
+using namespace cv;
+using namespace std;
+
+string hot_keys =
+    "\n\nHot keys: \n"
+    "\tESC - quit the program\n"
+    "\tq - quit the program\n"
+    "\tc - make the circle\n"
+    "\tr - make the rectangle\n"
+    "\th - make the convexhull\n"
+    "\tt - make the triangle\n"
+    "\te - make the ellipse\n"
+    "\ta - make all shapes\n";
+
+static void help(char** argv)
+{
+    cout << "\nThis program demonstrates various shape fitting techniques on a set of points using functions: \n"
+         << "minAreaRect(), minEnclosingTriangle(), minEnclosingCircle(), convexHull(), and ellipse().\n"
+         << "Three methods are used to find the elliptical fits: \n"
+         << "0: fitEllipse (OpenCV), 1: fitEllipseAMS, 2: fitEllipseDirect.\n"
+         << "These can be changed using the trackbar or command line argument.\n\n"
+         << "Usage: " << argv[0] << " [--image_name=<image_path> Default: ellipses.jpg] [--ellipse_method=<0/1/2> -- Default: 0 (OpenCV)]\n\n";
+    cout << hot_keys << endl;
+}
+
+void processImage(int, void*);
+void drawShapes(Mat &img, vector<Point> &points);
+void drawConvexHull(Mat &img, vector<Point> &points);
+void drawMinAreaRect(Mat &img, vector<Point> &points);
+void drawFittedEllipses(Mat &img, vector<Point> &points);
+void drawMinEnclosingCircle(Mat &img, vector<Point> &points);
+void drawMinEnclosingTriangle(Mat &img, const vector<Point> &points);
+
+// Shape fitting options
+Mat image;
+int sliderPos = 70;
+string shape = "--all";
+int ellipseMethod = 0;
```

Comment:
I propose to pass the value as function parameter in `drawFittedEllipses` and `drawShapes`

Reply:
Done by using userData struct